cmd/expiration-mailer/main.go

```diff
@@ -44,16 +44,75 @@ type regStore interface {
 	GetRegistration(ctx context.Context, req *sapb.RegistrationID, _ ...grpc.CallOption) (*corepb.Registration, error)
 }
 
+// limiter tracks how many mails we've sent to a given address in a given day.
+type limiter struct {
+	sync.RWMutex
+	// currentDay is a day in UTC, truncated to 24 hours. When the current
+	// time is more than 24 hours past this date, all counts reset and this
+	// date is updated.
+	currentDay time.Time
+
+	// counts is a map from address to number of mails we have attempted to
+	// send during `currentDay`.
+	counts map[string]int
+
+	// limit is the number of sends after which we'll return an error from
+	// check()
+	limit int
+
+	clk clock.Clock
+}
+
+// expects limiter is locked
+func (lim *limiter) maybeBumpDay() {
+	today := lim.clk.Now().UTC().Truncate(24 * time.Hour)
+	if (today.Sub(lim.currentDay) >= 24*time.Hour && len(lim.counts) > 0) ||
+		lim.counts == nil {
+		// Throw away counts so far and switch to a new day.
+		// This also does the initialization of counts and currentDay the first
+		// time inc() is called.
+		lim.counts = make(map[string]int)
+		lim.currentDay = today
+	}
+}
+
+// inc increments the count for the current day, and cleans up previous days
+// if needed.
+func (lim *limiter) inc(address string) {
+	lim.Lock()
+	defer lim.Unlock()
+
+	lim.maybeBumpDay()
+
+	lim.counts[address] += 1
+}
+
+// check checks whether the count for the given address is at the limit,
+// and returns an error if so.
+func (lim *limiter) check(address string) error {
```

Comment:
Should this be a tuple of `{accountID, address}`? If I had multiple ACME
accounts all configured with the same address, I would probably hit the
notification limit quickly. Then again, why would I really be watching my emails
for a heads up instead of relying on my clients automation.

Reply:
No; I suspect many of the integrators that receive excessive emails are using the same email address across multiple accounts. The natural unit of limitation for sending email is the recipient address, independent of the account.